src/windows.rs

```diff
@@ -2,16 +2,88 @@ use std::{
     convert::{TryFrom, TryInto},
     ffi::CStr,
     ffi::OsString,
+    net::IpAddr,
     os::windows::ffi::OsStringExt,
     ptr, slice,
 };
-use winapi::shared::{ntdef::ULONG, winerror::ERROR_SUCCESS, ws2def::AF_UNSPEC};
+use winapi::shared::{
+    ntdef::ULONG,
+    winerror::ERROR_SUCCESS,
+    ws2def::{AF_INET, AF_INET6, AF_UNSPEC, SOCKADDR_IN},
+    ws2ipdef::SOCKADDR_IN6,
+};
 use winapi::um::{iphlpapi::GetAdaptersAddresses, iptypes::IP_ADAPTER_ADDRESSES_LH};
 
 use crate::MacAddressError;
 
 const GAA_FLAG_NONE: ULONG = 0x0000;
 
+/// Similar to get_mac(); walk this list of adapters and in each
+/// adapter, walk the list of UnicastIpAddresses and return the mac address
+/// of the first one that matches the given IP
+pub(crate) fn get_mac_address_by_ip(ip: &IpAddr) -> Result<Option<[u8; 6]>, MacAddressError> {
+    let adapters = get_adapters()?;
+    // Safety: We don't use the pointer after `adapters` is dropped
+    let mut ptr = unsafe { adapters.ptr() };
+
+    // for each adapter on the machine
+    while !ptr.is_null() {
+        let bytes = unsafe { convert_mac_bytes(ptr) };
+
+        let mut ip_ptr = unsafe { (*ptr).FirstUnicastAddress };
+        // for each IP on the adapter
+        while !ip_ptr.is_null() {
+            let sock_addr = unsafe { (*ip_ptr).Address.lpSockaddr };
+            let adapter_ip = match unsafe { (*sock_addr).sa_family } as i32 {
+                AF_INET => unsafe {
+                    let addr = (*(sock_addr as *mut SOCKADDR_IN)).sin_addr;
+                    Some(IpAddr::from([
+                        addr.S_un.S_un_b().s_b1,
+                        addr.S_un.S_un_b().s_b2,
+                        addr.S_un.S_un_b().s_b3,
+                        addr.S_un.S_un_b().s_b4,
+                    ]))
+                },
+                AF_INET6 => unsafe {
+                    let addr = (*(sock_addr as *mut SOCKADDR_IN6)).sin6_addr;
+                    Some(IpAddr::from(addr.u.Byte().clone()))
+                },
+                _ => {
+                    // ignore unknown address families; only support IPv4/IPv6
+                    None
+                }
+            };
```

Comment:
I think `continue`ing here and advancing the pointer would again be cleaner so that you don't need to continually check for `Some`s. I'll have to review the rest of the code more closely on a Windows machine as well.

Reply:
Ok - I agree.  I can fix this up.  Let me know about the above and I'll change it all at once.

Reply:
I think that should solve my concerns for the time being then with the above feedback, I'll punt on the `HashMap` stuff for the time being (I want to do a bigger overhaul of the entire crate here at some point anyway so I'll investigate that closer then), so just let me know whenever things are updated and I'll merge :+1: